appserver/payara-appserver-modules/microprofile/metrics/src/main/java/fish/payara/microprofile/metrics/impl/MetricsConfigParserUtil.java

```diff
@@ -0,0 +1,180 @@
+                customPercentile = new MetricsCustomPercentiles(metricName, percentileValues);
+            }
+            metricPercentileCollection.addFirst(customPercentile);
+        }
+        return metricPercentileCollection;
+    }
+
+    public static Collection<MetricsCustomBuckets> parseHistogramBuckets(String histogramBucketsProperty) {
+        ArrayDeque<MetricsCustomBuckets> metricHistogramCollection = new ArrayDeque<>();
+        if (histogramBucketsProperty == null || histogramBucketsProperty.length() == 0) {
+            return null;
+        }
+
+        MetricsCustomBuckets customBucket = null;
+        String[] valuePairs = histogramBucketsProperty.split(PROPERTY_NAME_SEPARATOR);
+        for (String nameValue : valuePairs) {
+            String[] resultKeyValueSplit = nameValue.split(PROPERTY_KEY_VALUE_SEPARATOR);
+            String metricName = resultKeyValueSplit[0];
+            if (resultKeyValueSplit.length == 1) {
+                //evaluate when no value, disabled
+            } else {
+                Double[] bucketValues = Arrays.asList(resultKeyValueSplit[1].split(PROPERTY_VALUE_SEPARATOR)).stream()
+                        .map(MetricsConfigParserUtil::evaluateHistogramBucketValue)
+                        .filter(d -> d != null).toArray(Double[]::new);
+                Arrays.sort(bucketValues);
+                customBucket = new MetricsCustomBuckets(metricName, bucketValues);
+            }
+            metricHistogramCollection.addFirst(customBucket);
+        }
+        return metricHistogramCollection;
+    }
+
+    public static Collection<MetricsCustomBuckets> parseTimerBuckets(String timerBucketsProperty) {
+        ArrayDeque<MetricsCustomBuckets> metricTimerCollection = new ArrayDeque<>();
+        if (timerBucketsProperty == null || timerBucketsProperty.length() == 0) {
+            return null;
+        }
+
+        MetricsCustomBuckets customBucket = null;
+        String[] valuePairs = timerBucketsProperty.split(PROPERTY_NAME_SEPARATOR);
+        for (String nameValue : valuePairs) {
+            String[] resultKeyValueSplit = nameValue.split(PROPERTY_KEY_VALUE_SEPARATOR);
+            String metricName = resultKeyValueSplit[0];
+            if (resultKeyValueSplit.length != 1) {
+                Duration[] bucketValues = Arrays.asList(resultKeyValueSplit[1].split(PROPERTY_VALUE_SEPARATOR)).stream()
+                        .map(MetricsConfigParserUtil::evaluateTimerBucketValue)
+                        .filter(d -> d != null).toArray(Duration[]::new);
+                customBucket = new MetricsCustomBuckets(metricName,
+                        Arrays.stream(bucketValues).mapToDouble(Duration::toNanos).boxed().toArray(Double[]::new));
+            }
+            metricTimerCollection.addFirst(customBucket);
+        }
+        return metricTimerCollection;
+    }
+
+    public static synchronized Collection<MetricsCustomPercentiles> processPercentileMap(String appName) {
```

Comment:
Does this _need_ to be synchronized?

Is it an issue if there's some racing with resolving the config value - would it actually change that frequently?
Or is there some server startup or application loading race?

Reply:
same as before